### PR TITLE
feat: robust handling of wrapped arrays of reactant arrays

### LINKED ISSUE #131: Linear algebra on `TracedRArray`
The function `transpose` doesn't seem to work during Reactant compilation (it may not be the only one of its kind):

```julia
julia> using Reactant, LinearAlgebra

julia> r = Reactant.to_rarray([1.0, 2.0]);

julia> f(x) = sum(x .* transpose(x))
f (generic function with 1 method)

julia> f(r)
9.0

julia> @compile f(r)
ERROR: conversion to pointer not defined for Reactant.TracedRArray{Float64, 1}
Stacktrace:
  [1] error(s::String)
    @ Base ./error.jl:35
  [2] unsafe_convert(::Type{Ptr{Float64}}, a::Reactant.TracedRArray{Float64, 1})
    @ Base ./pointer.jl:67
  [3] unsafe_convert(::Type{Ptr{Float64}}, A::Transpose{Float64, Reactant.TracedRArray{Float64, 1}})
    @ LinearAlgebra ~/.julia/juliaup/julia-1.10.5+0.x64.linux.gnu/share/julia/stdlib/v1.10/LinearAlgebra/src/adjtrans.jl:348
  [4] pointer(x::Transpose{Float64, Reactant.TracedRArray{Float64, 1}})
    @ Base ./abstractarray.jl:1237
  [5] DenseElementsAttribute(values::Transpose{Float64, Reactant.TracedRArray{Float64, 1}})
    @ Reactant.MLIR.IR ~/.julia/packages/Reactant/c5fQ0/src/mlir/IR/Attribute.jl:553
  [6] broadcast_to_size(arg::Transpose{…}, rsize::Base.Generator{…})
    @ Reactant ~/.julia/packages/Reactant/c5fQ0/src/TracedRArray.jl:740
  [7] broadcast_to_size(arg::Base.Broadcast.Extruded{Transpose{…}, Tuple{…}, Tuple{…}}, rsize::Tuple{Int64, Int64})
    @ Reactant ~/.julia/packages/Reactant/c5fQ0/src/TracedRArray.jl:774
  [8] (::Reactant.var"#96#97"{Base.Broadcast.Broadcasted{…}})(a::Base.Broadcast.Extruded{Transpose{…}, Tuple{…}, Tuple{…}})
    @ Reactant ./none:0
  [9] iterate(g::Base.Generator{Tuple{…}, Reactant.var"#96#97"{…}}, s::Int64)
    @ Base ./generator.jl:47
 [10] _copyto!(dest::Reactant.TracedRArray{…}, bc::Base.Broadcast.Broadcasted{…})
    @ Reactant ~/.julia/packages/Reactant/c5fQ0/src/TracedRArray.jl:816
 [11] copyto!
    @ ~/.julia/packages/Reactant/c5fQ0/src/TracedRArray.jl:732 [inlined]
 [12] copyto!
    @ ./broadcast.jl:956 [inlined]
 [13] copy
    @ ~/.julia/packages/Reactant/c5fQ0/src/TracedRArray.jl:723 [inlined]
 [14] materialize
    @ ./broadcast.jl:903 [inlined]
 [15] f
    @ ./REPL[10]:1 [inlined]
 [16] (::Tuple{})(none::Reactant.TracedRArray{Float64, 1})
    @ Base.Experimental ./<missing>:0
 [17] (::Reactant.var"#26#35"{typeof(f), Reactant.MLIR.IR.Block, Vector{…}, Tuple{…}})()
    @ Reactant ~/.julia/packages/Reactant/c5fQ0/src/utils.jl:100
 [18] block!(f::Reactant.var"#26#35"{typeof(f), Reactant.MLIR.IR.Block, Vector{…}, Tuple{…}}, blk::Reactant.MLIR.IR.Block)
    @ Reactant.MLIR.IR ~/.julia/packages/Reactant/c5fQ0/src/mlir/IR/Block.jl:201
 [19] make_mlir_fn(f::Function, args::Tuple{…}, kwargs::Tuple{}, name::String, concretein::Bool; toscalar::Bool)
    @ Reactant ~/.julia/packages/Reactant/c5fQ0/src/utils.jl:74
 [20] make_mlir_fn
    @ ~/.julia/packages/Reactant/c5fQ0/src/utils.jl:24 [inlined]
 [21] #6
    @ ~/.julia/packages/Reactant/c5fQ0/src/Compiler.jl:253 [inlined]
 [22] block!(f::Reactant.Compiler.var"#6#11"{typeof(f), Tuple{ConcreteRArray{Float64, 1}}}, blk::Reactant.MLIR.IR.Block)
    @ Reactant.MLIR.IR ~/.julia/packages/Reactant/c5fQ0/src/mlir/IR/Block.jl:201
 [23] #5
    @ ~/.julia/packages/Reactant/c5fQ0/src/Compiler.jl:252 [inlined]
 [24] mmodule!(f::Reactant.Compiler.var"#5#10"{Reactant.MLIR.IR.Module, typeof(f), Tuple{…}}, blk::Reactant.MLIR.IR.Module)
    @ Reactant.MLIR.IR ~/.julia/packages/Reactant/c5fQ0/src/mlir/IR/Module.jl:93
 [25] compile_mlir!(mod::Reactant.MLIR.IR.Module, f::Function, args::Tuple{ConcreteRArray{Float64, 1}}; optimize::Bool)
    @ Reactant.Compiler ~/.julia/packages/Reactant/c5fQ0/src/Compiler.jl:249
 [26] compile_mlir!
    @ ~/.julia/packages/Reactant/c5fQ0/src/Compiler.jl:248 [inlined]
 [27] (::Reactant.Compiler.var"#30#32"{typeof(f), Tuple{ConcreteRArray{Float64, 1}}})()
    @ Reactant.Compiler ~/.julia/packages/Reactant/c5fQ0/src/Compiler.jl:576
 [28] context!(f::Reactant.Compiler.var"#30#32"{typeof(f), Tuple{ConcreteRArray{Float64, 1}}}, ctx::Reactant.MLIR.IR.Context)
    @ Reactant.MLIR.IR ~/.julia/packages/Reactant/c5fQ0/src/mlir/IR/Context.jl:71
 [29] compile_xla(f::Function, args::Tuple{ConcreteRArray{Float64, 1}}; client::Nothing)
    @ Reactant.Compiler ~/.julia/packages/Reactant/c5fQ0/src/Compiler.jl:573
 [30] compile_xla
    @ ~/.julia/packages/Reactant/c5fQ0/src/Compiler.jl:567 [inlined]
 [31] compile(f::Function, args::Tuple{ConcreteRArray{Float64, 1}}; client::Nothing)
    @ Reactant.Compiler ~/.julia/packages/Reactant/c5fQ0/src/Compiler.jl:600
 [32] compile(f::Function, args::Tuple{ConcreteRArray{Float64, 1}})
    @ Reactant.Compiler ~/.julia/packages/Reactant/c5fQ0/src/Compiler.jl:599
 [33] macro expansion
    @ ~/.julia/packages/Reactant/c5fQ0/src/Compiler.jl:360 [inlined]
 [34] top-level scope
    @ REPL[11]:1
Some type information was truncated. Use `show(err)` to see complete types.
```

### PR DESCRIPTION
**Pull request by avik-pal**
fixes #131

### CODE REVIEW

**Comment by mofeing** [comment]
I'm missing context here, probably because I missed last meeting 😅 

Why do we need this new `WrappedArray`? And we are we making `Adapt` a strong dependency?

**Comment by wsmoses** [comment]
So there are wrapper types in Julia. For example transpose(x) doesn’t actually return a new array but a view into x which is transposed. While we arguably could change that here the bigger issue is for a subarray. We want to keep the usual subarray structure and use those overloads.

However now subarray of rarray no longer triggers our overloads instead falling back to default implementations

**Comment by avik-pal** [comment]
Why do we define `Base.min`/`Base.max` on 2 arrays?

```julia
        function $jlop(
            @nospecialize(lhs::TracedRArray{T,N}), @nospecialize(rhs::TracedRArray{T2,N})
        ) where {T,T2,N}
            commonTy = TracedRArray{Base.promote_type(T, T2),N}
            lhs = promote_to(commonTy, lhs)
            rhs = promote_to(commonTy, rhs)
            return commonTy(
                (),
                MLIR.IR.result(
                    MLIR.Dialects.stablehlo.$hloop(lhs.mlir_data, rhs.mlir_data), 1
                ),
                size(lhs),
            )
        end
```

I don't think there is an equivalent Base version.

**Comment by wsmoses** [comment]
I don’t think we had the promote infrastructure yet.

If it’s fine without for one arg float32 one float64 using base promotes
then let’s clean it up!

File: src/TracedRArray.jl [hunk 1]
```diff
@@ -62,14 +69,19 @@ function Base.getindex(
...
-) where {T,N}
-    # TODO: Implement before merging the PR
-    return error("view is not supported yet")
+# Prevents ambiguity
+function Base.getindex(a::SubArray{T,N,<:AnyTracedRArray{T,N}}, indices::Int...) where {T,N}
+    return getindex(parent(a), Base.reindex(a.indices, indices)...)
+end
+
+function Base.getindex(a::SubArray{T,N,<:AnyTracedRArray{T,N}}, indices...) where {T,N}
+    return getindex(parent(a), Base.reindex(a.indices, indices)...)
```
Comment (avik-pal) [inline src/TracedRArray.jl:84]: Any idea if there is a general version of `reindex` that can take arbitrary arrays and compute the indices of the original unwrapped array.

Comment (wsmoses) [reply src/TracedRArray.jl:84]: Personally no, @vchuravy or @gbaraldi might have an idea?

**Comment by avik-pal** [comment]
We should have decent coverage of operations now. For now, it is mostly done for SubArrays but we can handle other wrappers by adding a dispatch to `get_ancestor_indices`

**Comment by wsmoses** [comment]
👍 honestly I think subarrays are the most critical atm anyways

**Review by wsmoses — APPROVED** [review]

**Pull request merged**